File: src/main/java/com/flab/infrun/member/infrastructure/jwt/TokenProvider.java
package com.flab.infrun.member.infrastructure.jwt;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import java.security.Key;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.stereotype.Component;

@Component
public class TokenProvider implements InitializingBean {

    private static final String AUTHORITIES_KEY = "auth";
    private final String secret;
    private final long tokenValidityInMilliseconds;
    private Key key;

    public TokenProvider(
        @Value("${jwt.secret}")
        String secret,
        @Value("${jwt.token-validity-in-seconds}")
        long tokenValidityInMilliseconds) {
        this.secret = secret;
        this.tokenValidityInMilliseconds = tokenValidityInMilliseconds * 1000;
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        final byte[] keyBytes = Decoders.BASE64.decode(secret);
        this.key = Keys.hmacShaKeyFor(keyBytes);
    }

    public String generateToken(Authentication authentication) {
        final String authorities = getAuthoritiesToString(authentication);
        final Date validity = getValdityDate();

        return Jwts.builder()
            .setSubject(authentication.getName())
            .claim(AUTHORITIES_KEY, authorities)
            .signWith(key, SignatureAlgorithm.HS256)
            .setExpiration(validity)
            .compact();
    }

    private String getAuthoritiesToString(final Authentication authentication) {
        return authentication.getAuthorities().stream()
            .map(GrantedAuthority::getAuthority)
            .collect(Collectors.joining(","));
    }

    private Date getValdityDate() {
        final long now = new Date().getTime();

        return new Date(now + this.tokenValidityInMilliseconds);
    }

    public Authentication getAuthentication(final String token) {
        final Claims claims = Jwts.parserBuilder()
            .setSigningKey(key)
            .build()
            .parseClaimsJws(token)
            .getBody();

        final List<SimpleGrantedAuthority> authorities = Arrays.stream(
                claims.get(AUTHORITIES_KEY).toString().split(","))
            .map(SimpleGrantedAuthority::new)
            .toList();

        final User principal = new User(claims.getSubject(), "", authorities);

        return new UsernamePasswordAuthenticationToken(principal, token, authorities);
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token);

            return true;
        } catch (Exception e) {
            return false;
        }
    }
}

File: src/main/kotlin/com/flab/infrun/member/infra/jwt/TokenProvider.kt
package com.flab.infrun.member.infra.jwt

import com.flab.infrun.common.domain.AuthenticatedUser
import com.flab.infrun.common.domain.Role
import com.flab.infrun.member.infra.properties.JwtProperties
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import java.security.Key
import java.time.Instant
import java.util.*

class TokenProvider(
    private val jwtProperties: JwtProperties,
) {

    private val keyBytes = Decoders.BASE64.decode(jwtProperties.secret)
    private var key: Key = Keys.hmacShaKeyFor(keyBytes)

    fun createToken(
        userId: Long,
        email: String,
        roles: List<Role>,
    ): String {
        val now = Instant.now()

        return encode(
            AuthenticatedUser(
                jti = UUID.randomUUID().toString(),
                userId = userId,
                email = email,
                roles = roles,
                issuer = jwtProperties.issuer,
                issuedAt = now,
                expiry = now.plusSeconds(jwtProperties.accessTokenExpirySecs),
            )
        )
    }

    private fun encode(authenticatedUser: AuthenticatedUser): String {
        return Jwts.builder()
            .setId(UUID.randomUUID().toString())
            .setSubject(authenticatedUser.userId.toString())
            .setIssuer(jwtProperties.issuer)
            .setIssuedAt(Date.from(Instant.now()))
            .claim("email", authenticatedUser.email)
            .claim("roles", authenticatedUser.roles)
            .signWith(key, SignatureAlgorithm.HS512)
            .setExpiration(Date.from(authenticatedUser.expiry))
            .compact()
    }

    fun decode(token: String): AuthenticatedUser {
        val claims = try {
            Jwts.parserBuilder()
                .setSigningKey(key)
                .requireIssuer(jwtProperties.issuer)
                .build()
                .parseClaimsJws(token)
                .body
        } catch (e: ExpiredJwtException) {
            e.claims
        } catch (e: Exception) {
            throw e
        }

        val jti = claims.id ?: throw MalformedJwtException("invalid jwt")
        val userId = claims.subject.toLong() ?: throw MalformedJwtException("invalid userId")
        val email =
            claims.get("email", String::class.java) ?: throw MalformedJwtException("invalid email")
        val roles = claims.get("roles", List::class.java).map { Role.valueOf(it as String) }
            ?: throw MalformedJwtException("invalid roles")
        val issuer = claims.issuer ?: throw MalformedJwtException("invalid issuer")
        val issuedAt = claims.issuedAt ?: throw MalformedJwtException("invalid issuedAt")
        val expiry = claims.expiration ?: throw MalformedJwtException("invalid expiry")

        return AuthenticatedUser(
            jti = jti,
            userId = userId,
            email = email,
            roles = roles,
            issuer = issuer,
            issuedAt = issuedAt.toInstant(),
            expiry = expiry.toInstant(),
        )
    }
}
